Fix max combo calculation in osu diffcalc and add tests

hlysine opened the pull request:
# Problem

https://github.com/ppy/osu/blob/ccac7b85be32b04a2297a9af045aa1ac4c022bec/osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs#L65-L67

The osu difficulty calculator did not consider the effect of classic mod when calculating max combo. Instead, it assumes that the max combo of a slider is simply equal to the number of nested hit objects in the slider, which is not the case if classic mod is not turned on.

**Example:**

For a slider with 2 slider ticks and no repeats, the hit results generated by the slider with classic mod turned on would be:
```
LargeTickHit (by slider head circle, combo +1)
LargeTickHit (by slider tick, combo +1)
LargeTickHit (by slider tick, combo +1)
SmallTickHit (by slider tail circle)
Great (by slider, combo +1)
```
Resulting in a max combo of 4, matching the number of nested hit objects in the slider.

However, if the classic mod isn't selected, the hit results become:
```
Great (by slider head circle, combo +1)
LargeTickHit (by slider tick, combo +1)
LargeTickHit (by slider tick, combo +1)
SmallTickHit (by slider tail circle)
IgnoreHit (by slider)
```
Resulting in a max combo of 3.

# Solution

~Osu classic mod sets `Slider.OnlyJudgeNestedObjects` to change the judgement behavior of the slider. Simply check for this boolean and subtract an additional 1 in the osu diffcalc to correct for max combo changes.~
Go over all hit objects and nested hit objects and check if the hit result affects combo, allowing effects of classic mod to be considered.

Tests are added for max combo calculation in all rulesets.

hlysine committed:
Fix max combo calculation in osu diffcalc

hlysine committed:
Add tests for diffcalc max combo

smoogipoo commented on osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs:
```diff
@@ -63,8 +63,10 @@ protected override DifficultyAttributes CreateDifficultyAttributes(IBeatmap beat
             int maxCombo = beatmap.HitObjects.Count;
-            // Add the ticks + tail of the slider. 1 is subtracted because the head circle would be counted twice (once for the slider itself in the line above)
-            maxCombo += beatmap.HitObjects.OfType<Slider>().Sum(s => s.NestedHitObjects.Count - 1);
+            // Add the ticks + tail of the slider
+            // 1 is subtracted because the head circle would be counted twice (once for the slider itself in the line above)
+            // an additional 1 is subtracted if only nested objects are judged because the hit result of the entire slider would not contribute to combo
+            maxCombo += beatmap.HitObjects.OfType<Slider>().Sum(s => s.NestedHitObjects.Count - 1 - (s.OnlyJudgeNestedObjects ? 1 : 0));
```
Better than this is to go over all hitobjects and their nested objects and look for judgements where `judgement => judgement.MaxResult.AffectsCombo()`

smoogipoo reviewed (dismissed):
This is a good change but before we do this we need to consider the implications of server-side diffcalc (i.e. attach classic mod). More of a reminder for myself than anything.

hlysine committed:
Simplify combo counting logic

smoogipoo committed:
Reimplement as extension method on IBeatmap

Implementation has changed slightly to support arbitrary levels of
nested hitobjects.

smoogipoo committed:
wip

smoogipoo committed:
Merge branch 'master' into osu-diff-calc-max-combo

smoogipoo committed:
Add remark about usage of CalculateAll()

smoogipoo commented:
Whops, didn't realise I had a 'wip' commit in this, but I can't force push :(

I initially went down a separate path to make `CreateDifficultyAdjustmentModCombinations()` properly handle the case where the classic mod is present across all mod combinations, resulting in duplicate mods which are not allowed by the current method. This is a non-trivial change which also meant rewriting the method to be iterative rather than recursive.

Instead of doing that, I came to the conclusion that adding the classic mod after-the-fact directly in `CalculateAll()` is a better way of going about this. I'm interested in others' thoughts on this path.

I've tested across all rulesets on all ranked beatmaps and there's 0 diffs with this PR as is.
- No changes are required to `osu-difficulty-calculator`. 
- Changes will be required for `osu-tools` and `osu-beatmap-difficulty-lookup-cache` in order to apply the classic mod to scores.

bdach commented on osu.Game/Rulesets/Difficulty/DifficultyCalculator.cs:
```diff
@@ -119,15 +120,23 @@ public List<TimedDifficultyAttributes> CalculateTimed([NotNull] IEnumerable<Mod>
         /// Calculates the difficulty of the beatmap using all mod combinations applicable to the beatmap.
         /// </summary>
+        /// <remarks>
+        /// This should only be used to compute difficulties for legacy mod combinations.
```
Just to confirm my understanding - due to the overall direction of lazer (no more ranked mods => number of combination increases exponentially) this method is mostly soft-deprecated anyway and shouldn't be used for anything serious going forward, right? In that case I think this solution is fine and probably even more correct than the initial one you gave, although it's not entirely clear to me what that one would entail from the description of that.

I'd maybe even propose renaming to `CalculateAllLegacyCombinations()` or something. It's a bit annoying  to make happen as it'll be a breaking change, but I think that name yells at whoever may want to latch onto this a bit more clearly that this is probably not the method they're looking for in the long run?

peppy replied:
I can get behind the rename if the intentions match as stated.

smoogipoo replied:
> Just to confirm my understanding

Your understanding is correct.

In the future we may use a function similar to this to compute a cache of the most common mod combinations/settings in order to not overload `osu-beatmap-difficulty-lookup-cache`, but I don't think it's going to be this method as it currently exists.

smoogipoo committed:
Merge branch 'master' into osu-diff-calc-max-combo

smoogipoo committed:
Rename method to CalculateAllLegacyCombinations()

bdach reviewed (approved):
Seems okay

peppy committed:
Merge branch 'master' into osu-diff-calc-max-combo
